### ISSUE TITLE
Device Expansion?

### ISSUE THREAD

**Issue by jaydiablo** [browscap/browscap#1239]
For #1235 I thought there was a lot of cut/paste to add the different devices I was adding for each agent, has there been any discussion/attempts of doing a Device expansion similar to how the platform expansion is done?

For example, most of the patterns in that PR were like these:

`Mozilla\/5.0 (#PLATFORM#HUAWEI Y336-U02 Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*`
`Mozilla\/5.0 (#PLATFORM#SM-G7102 Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*`
`Mozilla\/5.0 (#PLATFORM#HM NOTE 1S Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*`

Where the only thing that differs is the device identifier after the `#PLATFORM#` token.

Would it be possible to condense these to a single pattern, i.e.:

`Mozilla\/5.0 (#PLATFORM##DEVICE# Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*`

And then define the devices in a JSON array similar to how the platforms are done now?

```json
        {
          "match": "Mozilla\/5.0 (#PLATFORM##DEVICE# Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*",
          "engine": "U3",
          "devices": [
            "Xiaomi HM NOTE 1S",
            "Advan T1X Plus",
            "iPro Sense Golly"
          ],
          "platforms": [
            "Android_4_4"
          ]
        },
```

It's probably still wise to group devices that only use the same platforms (may help prevent fake useragents being detected, like if a device is known to not have android 6+ available for it), but even doing that would reduce the repetition quite a bit.

The devices would need a match property added, similar to the platforms:

```json
    "Xiaomi HM NOTE 1S": {
      "match": "HM NOTE 1S",
      "standard": false,
      "properties": {
        "Device_Type": "Mobile Phone",
        "Device_Name": "Redmi Note",
        "Device_Code_Name": "HM NOTE 1S",
        "Device_Maker": "Xiaomi Tech",
        "Device_Pointing_Method": "touchscreen",
        "Device_Brand_Name": "Xiaomi",
        "isMobileDevice": "true",
        "isTablet": "false"
      }
    },
```

Are there complexities that I'm overlooking?  I've noticed that Browscap lacks a little bit in the device detection, would this make it easier to add devices to different browsers?  Does it just blow up the number of patterns for "little" gain?

**Comment by mimmi20** [comment]
@jaydiablo Devices may occur with different writings.
For example the device `HTC DESIRE HD` (see #1219) also exists without `HTC` and with spaces replaced by underscores. In other devices the spaces may be replaced by slashes. Sometimes the same device exists with different names or occurs with the name of the vendor. I dont know if this was a issue only in older Android versions, or if this occurs new too.

This could be implemented with multiple matches, but I dont want to output each different writing for each device.

**Comment by jaydiablo** [comment]
Could there potentially be a `matches` array rather than just a `match` property that would define different matches that the device uses, then perhaps a syntax to specify which match should be used for that browser?

Maybe something like this for devices.json:

```json
"Xiaomi HM NOTE 1S": {
      "matches": {
        "full": "HM NOTE 1S",
        "short": "NOTE 1S",
        "arbitrary": "Xiamoi Note 1S"
      },
      "standard": false,
      "properties": {
        "Device_Type": "Mobile Phone",
        "Device_Name": "Redmi Note",
        "Device_Code_Name": "HM NOTE 1S",
        "Device_Maker": "Xiaomi Tech",
        "Device_Pointing_Method": "touchscreen",
        "Device_Brand_Name": "Xiaomi",
        "isMobileDevice": "true",
        "isTablet": "false"
      }
    },
```

`full` and `short` there don't have to be that way for all devices, but perhaps a naming convention makes sense.

Then in the applicable user-agent file, maybe something like this?

```json
        {
          "match": "Mozilla\/5.0 (#PLATFORM##DEVICE# Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*",
          "engine": "U3",
          "devices": {
            "Xiaomi HM NOTE 1S": "full",
            "Advan T1X Plus": "short",
            "iPro Sense Golly": "special"
          },
          "platforms": [
            "Android_4_4"
          ]
        },
```

The downside I suppose with that syntax is that it isn't possible to just define a device name without a match name (perhaps we could use something like "default" to mean the "0" index (or just the `match` property)).

Or, perhaps a compromise...   No changes to devices.json (no match/matches properties), but the match patterns can be specified in the devices array, maybe like so:

```json
        {
          "match": "Mozilla\/5.0 (#PLATFORM##DEVICE# Build\/*) AppleWebKit\/* (KHTML* like Gecko) Version\/* UCBrowser\/#MAJORVER#.#MINORVER#* U3\/* Safari\/*",
          "engine": "U3",
          "devices": {
            "Xiaomi HM NOTE 1S": "HM NOTE 1S",
            "Advan T1X Plus": "T1X Plus",
            "iPro Sense Golly": "Sense Golly"
          },
          "platforms": [
            "Android_4_4"
          ]
        },
```

This doesn't allow the device "patterns" to be re-used in the devices.json file at all, but maybe that's not that important, since, as you've seen, they're not terribly consistent from UA to UA.

This could reduce some of the cut & paste of the entire pattern though when all that changes is the device identification string.  Of course, when the INI file is built these would be expanded, replacing the `#DEVICE#` token with the value of each device (and assigning the device by name (from the key)).

**Comment by mimmi20** [comment]
I would prefer your last example.

**Comment by jaydiablo** [comment]
Here's a sample implementation of that last example:

https://github.com/jaydiablo/browscap/commit/9a2904ed8051fb80df142958c04b997d200649ff

Overall I think it could help consolidate quite a bit. It also may reveal patterns that differ slightly that may not need to.  For example, I didn't combine a couple because they differed by something like `*AppleWebKit*` vs. `*AppleWebKit\/*` even though everything else was the same (I didn't try changing one or the other and seeing if any tests failed, as that's sort of out of scope for this).

As mentioned in the commit, I didn't implement this for the parent patterns, just the children, but it may be possible to do for the parent as well if that seems useful (I suspect it's not, since there really is no consolidation to be done in the parent pattern).

Let me know what you think.

**Comment by mimmi20** [comment]
I would switch the keys and their values in the devices array. The string to replace should be the key in the array to be able to have the same device with different replacement string in one place.

for eaxmple:

```json
"devices": {
            "HTC DESIRE HD": "key1",
            "DESIRE HD": "key1",
            "HTC DESIRE": "key2"
          },
```

**Comment by jaydiablo** [comment]
Would it be less confusing to someone just reading the files if it was made more verbose, like so?

```json
"devices": [
  {"match": "HTC DESIRE HD", "name": "key1"},
  {"match": "DESIRE HD", "name": "key1"},
  {"match": "HTC DESIRE", "name": "key2"}
],
```

**Comment by jaydiablo** [comment]
My branch (https://github.com/jaydiablo/browscap/commits/device-expansion) now has the keys/values flipped as you suggested @mimmi20.  I also updated a couple more files. I actually did run into a couple of devices that wouldn't have worked if they were the other way around:

https://github.com/jaydiablo/browscap/commit/126f98d417079f8f5fd3775c169bece94325aac6#diff-4f47af8647cbbcc15461f1715803e42fR33
https://github.com/jaydiablo/browscap/commit/126f98d417079f8f5fd3775c169bece94325aac6#diff-4f47af8647cbbcc15461f1715803e42fR83

I'm open to moving to the more verbose version if that seems better.  I won't bother converting any more files over though until that's decided. :)
